### PR TITLE
[KYUUBI #3851][SPARK] Support auto set up `spark.master` when Kyuubi running inside Pod

### PR DESCRIPTION
**Pull request by xuefeimiaoao**
### _Why are the changes needed?_

to close https://github.com/apache/incubator-kyuubi/issues/3851 .
We can deploy spark on kubernetes without configure spark.master explicitly when api-server url is not exposed for us.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x]  Add screenshots for manual tests if appropriate
we do not set `spark.master`:
企业微信截图_16692948916568

kyuubi add `spark.master` when setting `kyuubi.kubernetes.engine.master.override` to true.
企业微信截图_16692948189845

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:68]: how about try auto configure it iff `spark.master` is absent?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:72]: w/ hardcoded "https", how to support "http"?

**Comment by pan3793** [comment]
Thanks @xuefeimiaoao, it's good to make Kyuubi smart and out-of-box for Spark on K8s, left some comments.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: I agree with your view and complete `spark.master` only in case of  its absence.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: After checking "https://kubernetes.io/docs/concepts/architecture/control-plane-node-communication/#node-to-control-plane", the API server is configured to listen for remote connections on a secure HTTPS port (typically 443), so just set https here.

Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: If user want to use http, he could set `spark.master` explicitly, which we will respect and will not be overridden after changing of my code.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:71]: I think `clusterManager()` should be called instead

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:71]: And the code can be simplified to
```
(clusterManager(), sys.env.get(KUBERNETES_SERVICE_HOST), sys.env.get(KUBERNETES_SERVICE_PORT)) match {
  case (None, Some(kubernetesServiceHost), Some(kubernetesServicePort)) =>
    ...
  case _ =>
}
```

Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:71]: `      (
        clusterManager(),
        sys.env.get(KUBERNETES_SERVICE_HOST),
        sys.env.get(KUBERNETES_SERVICE_PORT)) match {
        case (None, Some(kubernetesServiceHost), Some(kubernetesServicePort)) =>
          // According to "https://kubernetes.io/docs/concepts/architecture/control-plane-
          // node-communication/#node-to-control-plane", the API server is configured to listen
          // for remote connections on a secure HTTPS port (typically 443), so we set https here.
          val masterURL = s"k8s://https://${kubernetesServiceHost}:${kubernetesServicePort}"
          conf.set(MASTER_KEY, masterURL)
        case (Some(clusterManager), _, _) =>
          // in case that ClusterManager was set in spark-defaults.conf, which should be set into
          // KyuubiConf, the one which is used to build sparkSubmit.
          conf.setIfMissing(MASTER_KEY, clusterManager)
        case _ =>
      }`

Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: I think `clusterManager()` is better.

Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: Besides, i found that `spark.master` will be lost if which is set under SPARK_CONF_DIR, so i add some code.

Comment (xuefeimiaoao) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: Besides, i found that spark.master will be lost if which is set under SPARK_CONF_DIR, which seems like a bug, so i add some code.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala
Comment (xuefeimiaoao) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:84]: How about complete spark.master if which is set into spark-defaults.conf and not in kyuubiConf, in which case spark.master will be lost when build sparkSubmit.

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:84]: it's redundant, please remove it and keep thing simple.

**Review by pan3793 — APPROVED** [review]
Thanks, LGTM, pending CI, and cc @hddong @zwangsheng

**Comment by zwangsheng** [comment]
In the case you mentioned that the users should not aware of the kubernetes api service address, should we consider config `spark.kubernetes.driver.master` for users in cluster mode?

**Comment by xuefeimiaoao** [comment]
> In the case you mentioned that the users should not aware of the kubernetes api service address, should we consider config `spark.kubernetes.driver.master` for users in cluster mode?
@zwangsheng
I suppose that we could respect user defined config, whether it is `spark.master` or `spark.kubernetes.driver.master`, nonetheless, we actually do not need those configs in most cases.

**Comment by zwangsheng** [comment]
> I suppose that we could respect user defined config, whether it is `spark.master` or `spark.kubernetes.driver.master`, nonetheless, we actually do not need those configs in most cases.

I understand what you said and this pr LGTM. I mean in the case we mentioned that we need to help the users set up `spark.master` (Users can't sense kubernetes master and commit using cluster deploy mode), do we need to help the user configure this `spark.kubernetes.driver.master` at the same time?

**Comment by xuefeimiaoao** [comment]
> 

I suppose that the default value of `spark.kubernetes.driver.master` in spark is just ok.

**Review by zwangsheng — APPROVED** [review]

**Comment by xuefeimiaoao** [comment]
CI has been blocked by test of DiscoveryClientTests, could you help to skip it or trigger recheck again?@zwangsheng

**Comment by zwangsheng** [comment]
> CI has been blocked by test of DiscoveryClientTests, could you help to skip it or trigger recheck again?@zwangsheng

@pan3793 ,ping. 

Need retry CI.

**Comment by pan3793** [comment]
"CI / Kyuubi Server On Kubernetes Integration Test" failed, and it's related

**Comment by xuefeimiaoao** [comment]
I am confused how `org.apache.kyuubi.kubernetes.test.deployment.KyuubiOnKubernetesWithLocalSparkTestsSuite#connectionConf` passed to kyuubi server when testing. @zwangsheng

**Comment by xuefeimiaoao** [comment]
> 

I suppose that spark will use default master `local[*]` encountering missing `spark.master` according to `org.apache.spark.deploy.SparkSubmitArguments#loadEnvironmentArguments`, which incurs a problem of `service account doesn't have access` cus serviceaccount was not set, so that we should skipping setting `spark.master` when both `spark.master` and `spark.submit.deployMode` are not set. In another word, we will just auto-set `spark.master` when `spark.master` is not set and `spark.submit.deployMode` is set explicitly.@pan3793

**Comment by pan3793** [comment]
> ... skipping setting spark.master when both spark.master and spark.submit.deployMode are not set.

sounds very limit.

> I am confused how `org.apache.kyuubi.kubernetes.test.deployment.KyuubiOnKubernetesWithLocalSparkTestsSuite#connectionConf` passed to kyuubi server when testing

It set `"spark.master" -> "local"`, should not be affected?

**Comment by zwangsheng** [comment]
> I am confused how `org.apache.kyuubi.kubernetes.test.deployment.KyuubiOnKubernetesWithLocalSparkTestsSuite#connectionConf` passed to kyuubi server when testing. @zwangsheng

It seems we are not using this connect conf, I'll turn on a pr to fix.

**Comment by xuefeimiaoao** [comment]
> > ... skipping setting spark.master when both spark.master and spark.submit.deployMode are not set.
> 
> sounds very limit.

Indeed, and could you help me review and retrigger ci again

**Comment by pan3793** [comment]
@xuefeimiaoao would mind send a simple pr to unlock the ci? e.g. https://github.com/apache/incubator-kyuubi/issues/3861 once a pr got merge, you don't need to approve to run ci

**Comment by pan3793** [comment]
@xuefeimiaoao I in favor the previous idea, we don't need to check the deploy mode. let's wait @zwangsheng to fix the `connectionConf` first

**Comment by xuefeimiaoao** [comment]
> @xuefeimiaoao would mind sending a simple pr to unlock the ci? e.g. #3861 once a pr get merged, you don't need to wait for approve to run ci

With pleasure

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**
